### DISCUSSION TITLE
Button to swap toolbars (courtesy of a Templater script)

### DISCUSSION THREAD

**Discussion by FeralFlora** [chrisgurney/obsidian-note-toolbar#79] [Scripting]
![Obsidian_z4d1Ui4afk](https://github.com/chrisgurney/obsidian-note-toolbar/assets/33608083/e887d38b-f539-437e-95fc-9529f9d0ab2a)

I've made a little Templater script that lets you easily swap toolbars. 

This works by retrieving all the toolbars' names directly from the Note toolbar settings, and presenting them in a suggester. 

After you pick a toolbar in the suggester, a `toolbarProp` property is either added or updated in the current note to reflect your choice. This swaps the current toolbar to your chosen toolbar. The `toolbarProp` is also taken from the plugin's settings, so you don't have to configure anything.

> [!warning]
> This button needs to modify the frontmatter of your current note in order to work. Of course, this only occurs when you swap a toolbar.

## The template

```js
<%*
// Check if anything is selected. If it is, return it, so it's not lost.
const selection = tp.file.selection();
if (selection.length > 0) {
	tR += selection;
}
tp.hooks.on_all_templates_executed(async () => {
	
	// Get all the configured toolbars and the toolbar property from the plugin settings
	const { toolbars, toolbarProp } = app.plugins.getPlugin("note-toolbar").settings;
	
	// Option to revert to default folder mapping
	const defaultOption = "Default folder mapping";
	
	// Extract only the toolbar names and add the default option
	const toolbarOptions = toolbars.map(toolbar => toolbar.name).concat(defaultOption);
	
	// Prompt the user to pick a toolbar
	const pickedToolbar = await tp.system.suggester((item) => item, toolbarOptions, true, "Swap to toolbar:");
	
	// Get the current file
	const currentFile = app.workspace.getActiveFile();

	// Update the frontmatter of the current file with the selected toolbar option
	await app.fileManager.processFrontMatter(currentFile, (frontmatter) => {
		
		// If the user picked the default option, delete the toolbar property
		if (pickedToolbar === defaultOption) {
			delete frontmatter[toolbarProp];
			return;
		}
		
		// Otherwise, set the toolbar property to the chosen toolbar
		frontmatter[toolbarProp] = pickedToolbar;
	});
});
_%>

```

## Setup

**Prerequisites**:
- Templater plugin

**Steps**:
1. Save the template code above into a template note in your Template folder, as configured in Templater's settings, and give it a descriptive name like `swapToolbar`
2. Bind the template to a command in Templater's settings
![image](https://github.com/chrisgurney/obsidian-note-toolbar/assets/33608083/cb83fa3c-01c0-420c-a50c-dfcab881b04e)
3. Add a "Swap toolbar" button to your toolbars, using the command type and then choose the command configured in step 2 as the command

Voilà! Now you can easily swap out toolbars for different workflows at different times.

The most involved part about this is that you have to configure the "Swap toolbar" button in each of the toolbars where you want it to be present. It would be nice to be able to re-use already configured items from existing toolbars, when setting up toolbar buttons / items.

## Forum post

https://forum.obsidian.md/t/note-toolbar-button-to-easily-swap-toolbars-courtesy-of-a-templater-script/82350?u=feralflora

**Comment by chrisgurney** [comment]
@FeralFlora Wow this is amazing!

**Reply by FeralFlora** [reply]
Thank you :) I've updated the code so that there is an option to revert to the default folder mapping (delete the toolbar property).

![Obsidian_116dYGa0oQ](https://github.com/chrisgurney/obsidian-note-toolbar/assets/33608083/526015b5-bf04-45fd-ad7c-a676f2eff412)

And I also fixed a whitespace issue that led to blank lines being inserted into the note.

**Reply by FeralFlora** [reply]
(and added comments)

**Comment by MrComputron** [comment]
@FeralFlora With the new menu item you can create a sub toolbar with just the swap Toolbar item all formatted and place that in your other toolbars. As it is a full toolbar (even if just one button) it maintains all the styling when inserted. Much less hassle than rebuilding that command each for each toolbar.

**Comment by chrisgurney** [comment]
Hi @FeralFlora I've added your helpful script to the [examples](https://github.com/chrisgurney/obsidian-note-toolbar/tree/master/examples/Scripts) folder, now that it can be executed directly using the Templater integration:

https://raw.githubusercontent.com/chrisgurney/obsidian-note-toolbar/refs/heads/master/examples/Scripts/Templater/SwapToolbars.js

I've commented out some of the lines as I don't think they're now needed in this context (I also wasn't sure what the `tR` variable was used for). It works for me as-is. Are you able to review / confirm when you have a moment?

**Reply by FeralFlora** [reply]
Hi Chris, that's nice!

Since you changed it into a user script, nothing is inserted as a template, so you're right that the selection return and `hooks.on_all_templates_executed` aren't needed anymore. I tested the user script, and it seems to work as expected 👍 

In my original implementation, the selection would be lost if it wasn't returned as part of `tR` ([which is expected behavior](https://github.com/SilentVoid13/Templater/issues/1058#issuecomment-1809502944)), the variable that Templater uses to store command results:

> When our templating engine generates a replacement string using all of our commands results, it is stored in a variable named tR. This is the string that will contain the processed file content. You are allowed to access that variable from a JS execution command. This means that, to output something from a JS execution command, you just need to append what you want to output to that tR string variable. 

https://silentvoid13.github.io/Templater/commands/execution-command.html#:~:text=When%20our%20templating,tR%20string%20variable.

So I think you can just delete the parts you commented out. Cheers!

**Reply by chrisgurney** [reply]
That's great, thank you so much! And thanks again for the script!

**Comment by chrisgurney** [comment]
@FeralFlora I finally came to the conclusion that this should have been added ages ago to the plugin. It's now available in [the beta](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/246), with the added benefit of showing toolbar previews in the suggester.

I am curious -- since I'm using your frontmatter approach -- if there should be any warnings in the UI (apart from a note in the User Guide) to indicate that the note is being, or will be, modified. What are your thoughts on that?

**Reply by chrisgurney** [reply]
I thought of one issue, where it's been suggested a couple times to users who want toolbars based on tags to change the property to be `tags`. Restoring the toolbar to default would be destructive in this case.

Solutions:
- Do not show the option of the prop is set to `tags` or 
- see if the toolbar name can be removed from the frontmatter prop string and maybe then only delete the prop if what's left is an empty string.

**Reply by FeralFlora** [reply]
That's great, thanks for letting me know and the credit <3 The toolbar previews are a great touch!

As for the question about warnings, I think there should be some little note somewhere in the modal. This is mainly because processFrontmatter changes the YAML (removes comments, changes array formatting etc.), so it's good to give people a heads up, if they want to avoid that. But perhaps the update changelog is enough (can't remember, but you have one in-app, right?)

**Reply by chrisgurney** [reply]
Thank you. I've added mention to the release notes, but thinking about it further, it won't help new users down the road.

I'm working on a generic dismissible onboarding message that should help cover this concern.

**Reply by chrisgurney** [reply]
A preview:

<img src="https://github.com/user-attachments/assets/b07624f8-f551-4f54-a478-522e9c5b1b74" width="400"/>
